pkg/kubelet/preemption/preemption.go

```diff
@@ -0,0 +1,264 @@
+	resourceReasons := []*admissionRequirement{}
+	for _, reason := range failureReasons {
+		if r, ok := reason.(*predicates.InsufficientResourceError); ok {
+			resourceReasons = append(resourceReasons, &admissionRequirement{
+				resourceName: r.ResourceName,
+				quantity:     r.GetInsufficientAmount(),
+			})
+		} else {
+			nonResourceReasons = append(nonResourceReasons, reason)
+		}
+	}
+	if len(nonResourceReasons) > 0 {
+		// Return only reasons that are not resource related, since critical pods cannot fail admission for resource reasons.
+		return false, nonResourceReasons, nil
+	}
+	err := c.evictPodsToFreeRequests(admissionRequirementList(resourceReasons))
+	// if no error is returned, preemption succeeded and the pod is safe to admit.
+	return err == nil, nil, err
+}
+
+// freeRequests takes a list of insufficient resources, and attempts to free them by evicting pods
+// based on requests.  For example, if the only insufficient resource is 200Mb of memory, this function could
+// evict a pod with request=250Mb.
+func (c *CriticalPodAdmissionHandler) evictPodsToFreeRequests(insufficientResources admissionRequirementList) error {
+	podsToPreempt, err := getPodsToPreempt(c.getPodsFunc(), insufficientResources)
+	if err != nil {
+		return fmt.Errorf("preemption: error finding a set of pods to preempt: %v", err)
+	}
+	glog.Infof("preemption: attempting to evict pods %v, in order to free up resources: %s", podsToPreempt, insufficientResources.toString())
+	for _, pod := range podsToPreempt {
+		status := v1.PodStatus{
+			Phase:   v1.PodFailed,
+			Message: message,
+			Reason:  events.PreemptContainer,
+		}
+		// record that we are evicting the pod
+		c.recorder.Eventf(pod, v1.EventTypeWarning, events.PreemptContainer, message)
+		// this is a blocking call and should only return when the pod and its containers are killed.
+		err := c.killPodFunc(pod, status, nil)
+		if err != nil {
+			return fmt.Errorf("preemption: pod %s failed to evict %v", format.Pod(pod), err)
+		}
+		glog.Infof("preemption: pod %s evicted successfully", format.Pod(pod))
+	}
+	return nil
+}
+
+// getPodsToPreempt returns a list of pods that could be preempted to free requests >= requirements
+func getPodsToPreempt(pods []*v1.Pod, requirements admissionRequirementList) ([]*v1.Pod, error) {
+	bestEffortPods, burstablePods, guaranteedPods := sortPodsByQOS(pods)
+
+	// make sure that pods exist to reclaim the requirements
+	unableToMeetRequirements := requirements.subtract(append(append(bestEffortPods, burstablePods...), guaranteedPods...)...)
+	if len(unableToMeetRequirements) > 0 {
+		return nil, fmt.Errorf("no set of running pods found to reclaim resources: %v", unableToMeetRequirements.toString())
+	}
+	// find the guaranteed pods we would need to evict if we already evicted ALL burstable and besteffort pods.
+	guarateedToEvict, err := getPodsToPreemptByDistance(guaranteedPods, requirements.subtract(append(bestEffortPods, burstablePods...)...))
```

Comment:
We dont know we will have to evict all burstable and besteffort.  That is the point of finding the guaranteed we **have to** evict first.  That way, if we have to evict a large, guaranteed pod, we at least avoid evicting a whole bunch of burstable pods as well, and only evict the ones we need to.  

Ill revisit this and add a comment for the first part after code freeze.